src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs

```diff
@@ -0,0 +1,1343 @@
+
+                    i2 <<= 6;
+
+                    i0 |= i3;
+                    i0 |= i2;
+
+                    if (i0 < 0)
+                    {
+                        goto InvalidDataExit;
+                    }
+                    if (dest + 3 > destMax)
+                    {
+                        goto DestinationTooSmallExit;
+                    }
+
+                    WriteThreeLowOrderBytes(dest, i0);
+                    dest += 3;
+                    src += 4;
+                }
+                else if (!IsValidPadding(t2))
+                {
+                    int i2 = Unsafe.Add(ref decodingMap, (IntPtr)t2);
+
+                    i2 <<= 6;
+
+                    i0 |= i2;
+
+                    if (i0 < 0)
+                    {
+                        goto InvalidDataExit;
+                    }
+                    if (dest + 2 > destMax)
+                    {
+                        goto DestinationTooSmallExit;
+                    }
+
+                    dest[0] = (byte)(i0 >> 16);
+                    dest[1] = (byte)(i0 >> 8);
+                    dest += 2;
+                    src += remaining;
+                }
+                else
+                {
+                    if (i0 < 0)
+                    {
+                        goto InvalidDataExit;
+                    }
+                    if (dest + 1 > destMax)
+                    {
+                        goto DestinationTooSmallExit;
+                    }
+
+                    dest[0] = (byte)(i0 >> 16);
+                    dest += 1;
+                    src += remaining;
+                }
+
+                if (srcLength != source.Length)
+                {
+                    goto InvalidDataExit;
+                }
+
+            DoneExit:
+                bytesConsumed = (int)(src - srcBytes);
+                bytesWritten = (int)(dest - destBytes);
+                return OperationStatus.Done;
+
+            DestinationTooSmallExit:
+                if (srcLength != source.Length && isFinalBlock)
+                {
+                    goto InvalidDataExit; // if input is not a multiple of 4, and there is no more data, return invalid data instead
+                }
+
+                bytesConsumed = (int)(src - srcBytes);
+                bytesWritten = (int)(dest - destBytes);
+                return OperationStatus.DestinationTooSmall;
+
+            NeedMoreDataExit:
+                bytesConsumed = (int)(src - srcBytes);
+                bytesWritten = (int)(dest - destBytes);
+                return OperationStatus.NeedMoreData;
+
+            InvalidDataExit:
+                bytesConsumed = (int)(src - srcBytes);
+                bytesWritten = (int)(dest - destBytes);
+                return ignoreWhiteSpace ?
+                    InvalidDataFallback(source, bytes, ref bytesConsumed, ref bytesWritten, isFinalBlock) :
+                    OperationStatus.InvalidData;
+            }
```

Comment:
There's a lot of code being duplicated here. Is there no opportunity for further code sharing with the .NET 9 implementation?

Reply:
Because cannot use static abstract methods in netstandard the main implementation cannot be shared, though I think some helper methods could be shared, assuming it will not need to add a netstandard target for CoreLib

Reply:
Would we be able to share it if we just used non-static abstract methods? e.g. instead have instance interface methods, implement those as instance methods on a struct, and pass in a default instance of the struct as an argument?

Reply:
I guess it could, but how that would affect performance?

Reply:
I'd hope very little, but you'd need to actually measure to see.

Reply:
OK, I'll give it try